File: Algorithms/198. House Robber/swift/solution.dp.swift
class Solution {
    func rob(_ nums: [Int]) -> Int {
        if nums.count == 0 {
            return 0
        }

        let N: Int = nums.count
        var maxRobbedAmount: [Int] = Array(repeating: 0, count: N + 1)
        maxRobbedAmount[N] = 0
        maxRobbedAmount[N - 1] = nums[N - 1]

        for i: Int in (0..<N - 1).reversed() {
            maxRobbedAmount[i] = max(maxRobbedAmount[i + 1], nums[i] + maxRobbedAmount[i + 2])
        }

        return maxRobbedAmount[0]
    }
}

File: Algorithms/198. House Robber/golang/solution.dp.go
func rob(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	var N int = len(nums)
	var maxRobbedAmount []int = make([]int, N+1)

	maxRobbedAmount[N], maxRobbedAmount[N-1] = 0, nums[N-1]
	for i := N - 2; i >= 0; i-- {
		if maxRobbedAmount[i+1] > maxRobbedAmount[i+2]+nums[i] {
			maxRobbedAmount[i] = maxRobbedAmount[i+1]
		} else {
			maxRobbedAmount[i] = maxRobbedAmount[i+2] + nums[i]
		}
	}
	return maxRobbedAmount[0]
}
